src/main/java/com/redhat/devtools/lsp4ij/features/formatting/LSPEditorImprovementsTypedHandler.java

```diff
@@ -0,0 +1,208 @@
+    }
+
+    private boolean handleStatementTerminator(@NotNull PsiFile file,
+                                              @NotNull Editor editor) {
+        Document document = editor.getDocument();
+        CaretModel caretModel = editor.getCaretModel();
+        int offset = caretModel.getOffset();
+        if (offset == document.getTextLength()) {
+            return false;
+        }
+
+        CharSequence documentChars = document.getCharsSequence();
+        char nextCharacter = documentChars.charAt(offset);
+        if (isStatementTerminatorCharacter(file, nextCharacter) &&
+            !inStringLiteral(file, documentChars, offset) &&
+            !inComment(file, documentChars, offset)) {
+            caretModel.moveToOffset(offset + 1);
+            return true;
+        }
+
+        return false;
+    }
+
+    @Nullable
+    private static String getStringLiteral(@NotNull PsiFile file,
+                                           @NotNull CharSequence fileChars,
+                                           int offset) {
+        Set<Character> quoteCharacters = LSPIJEditorUtils.getQuoteCharacters(file);
+        if (!ContainerUtil.isEmpty(quoteCharacters)) {
+            // This pattern is basically from any quote character to the same character with no unescaped same quote
+            // character or line break in between
+            Pattern stringLiteralPattern = Pattern.compile("([" + StringUtil.join(quoteCharacters, "") + "])([^$1\\r\\n]|\\\\\\.)*\\1");
+            Matcher stringLiteralMatcher = stringLiteralPattern.matcher(fileChars);
+            while (stringLiteralMatcher.find()) {
+                int stringLiteralStart = stringLiteralMatcher.start();
+                int stringLiteralEnd = stringLiteralMatcher.end();
+                if ((stringLiteralStart < offset) && (offset < stringLiteralEnd)) {
+                    return fileChars.subSequence(stringLiteralStart, stringLiteralEnd).toString();
+                }
+            }
+        }
+
+        return null;
+    }
+
+    private static boolean inStringLiteral(@NotNull PsiFile file,
+                                           @NotNull CharSequence documentChars,
+                                           int offset) {
+        return getStringLiteral(file, documentChars, offset) != null;
+    }
+
+    private static boolean inComment(@NotNull PsiFile file,
+                                     @NotNull CharSequence documentChars,
+                                     int offset) {
+        boolean inComment = false;
+
+        Document document = LSPIJUtils.getDocument(file);
+        if (document != null) {
+            Commenter commenter = LSPIJEditorUtils.getCommenter(file);
+
+            // First try to see if we're in a line comment since that's much more efficient. We're in a lne
+            // comment if we find a line comment start on the same line before the current position.
+            String lineCommentPrefix = commenter.getLineCommentPrefix();
```

Comment:
Line comment is easy. We just see if there's a line comment prefix before the caret in the current line.